packages/framework/ai-collab/src/explicit-strategy/agentEditReducer.ts

```diff
@@ -0,0 +1,570 @@
+						(parentNode as any)[fieldKey] = undefined;
+					} else {
+						throw new UsageError(
+							`${fieldKey} is required, and cannot be removed. Please use modify edit instead.`,
+						);
+					}
+				}
+			} else if (isRange(source)) {
+				const { array, startIndex, endIndex } = getRangeInfo(source, idGenerator);
+				array.removeRange(startIndex, endIndex);
+			}
+			return treeEdit;
+		}
+		case "modify": {
+			const node = getNodeFromTarget(treeEdit.target, idGenerator);
+			const { treeNodeSchema } = getSimpleNodeSchema(node);
+
+			const fieldSchema =
+				(treeNodeSchema.info as Record<string, ImplicitFieldSchema>)[treeEdit.field] ??
+				fail("Expected field schema");
+
+			const modification = treeEdit.modification;
+
+			// eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
+			const schemaIdentifier = (modification as any)[typeField];
+
+			let insertedObject: TreeNode | undefined;
+			// if fieldSchema is a LeafnodeSchema, we can check that it's a valid type and set the field.
+			if (isPrimitive(modification)) {
+				// eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-member-access
+				(node as any)[treeEdit.field] = modification;
+			}
+			// If the fieldSchema is a function we can grab the constructor and make an instance of that node.
+			else if (typeof fieldSchema === "function") {
+				const simpleSchema = fieldSchema as unknown as new (dummy: unknown) => TreeNode;
+				populateDefaults(modification, definitionMap);
+				const constructedModification = new simpleSchema(modification);
+				validator?.(constructedModification);
+				insertedObject = constructedModification;
+
+				if (Array.isArray(modification)) {
+					// eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-member-access
+					const field = (node as any)[treeEdit.field] as TreeArrayNode;
+					assert(Array.isArray(field), "the field must be an array node");
+					assert(
+						Array.isArray(constructedModification),
+						"the modification must be an array node",
+					);
+					field.removeRange(0);
+					// eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-member-access
+					(node as any)[treeEdit.field] = constructedModification;
+				} else {
+					// eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-member-access
+					(node as any)[treeEdit.field] = constructedModification;
+				}
+			}
+			// If the fieldSchema is of type FieldSchema, we can check its allowed types and set the field.
+			else if (fieldSchema instanceof FieldSchema) {
+				if (fieldSchema.kind === FieldKind.Optional && modification === undefined) {
+					// eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-member-access
+					(node as any)[treeEdit.field] = undefined;
+				} else {
```

Comment:
(placement of this comment is somewhat arbitrary)

@noencke @taylorsw04 we do a lot of manual schema validation in this file, which requires adding exports of functionality like `normalizeAllowedTypes`, which really shouldn't be exposed outside of the tree library.

Similar validation occurs when we apply the edit to the tree, doesn't it? Can we not just wrap our edits in try-catch blocks instead, and let natural validation take place, rather than re-implementing schema validation here?

Reply:
AFAIK, The validation that exists inside the tree is not enabled properly. I believe currently, the failure mode we would end up in is that we fail somewhere deep in the forest without a useful error message. In this case, the error message is actually critical as it is what we feedback into the model. So, until this issue is fixed, I think we have to go with the extra validation layer here. Also, note that a decent amount of the validation is for AI specific context such as ranges and their endpoints.